File: 630-course-schedule-iii/630-course-schedule-iii.cpp
class Solution {
public:
    static bool comp(const vector<int>& vec1, const vector<int>& vec2) {
        return vec1[1] < vec2[1];
    }
    int scheduleCourse(vector<vector<int>>& courses) {
        sort(courses.begin(), courses.end(), comp);  // sort by duration
        priority_queue<int> heap;  // it's by default a max heap!
        int total = 0;  // sum of courses length
        // iterate elements of courses
        int duration;
        int lastDay;
        for (int i=0; i<courses.size(); i++) {
            duration = courses[i][0];
            lastDay = courses[i][1];
            if ((total + duration) <= lastDay) {  // we can add this course
                total += duration;
                heap.push(duration);
            }
            else {
                if (heap.size() != 0) {
                    if (heap.top() > duration) {
                        total += (-heap.top() + duration);
                        heap.pop();
                        heap.push(duration);
                    }
                }
            }
        }
        return heap.size();
    }
};

File: 630-course-schedule-iii/630-course-schedule-iii.py
import heapq
class Solution:
    # I failed ...
    # Trick: sort by lastDay,
    def scheduleCourse(self, courses: List[List[int]]) -> int:
        heap = []
        heapq.heapify(heap)  # we will add negative values, it's a max heap
        total = 0  # total duration of the chosen courses
        courses.sort(key=lambda x: x[1])  # sort by lastDay
        for duration, lastDay in courses:
            if duration + total <= lastDay:  # we can take the course
                total += duration
                heapq.heappush(heap, -duration)
            elif heap and -heap[0] > duration:  # remove the longest course to fit this new one
                total += duration + heapq.heappop(heap)
                heapq.heappush(heap, -duration)
        return len(heap)
